test/Microsoft.VisualStudio.Threading.Tests/RarelyRemoveItemSetTests.cs

```diff
@@ -0,0 +1,141 @@
+            enumerator.Reset();
+            Assert.False(enumerator.MoveNext());
+        }
+    }
+
+    [Fact]
+    public void EnumerationOfOne()
+    {
+        this.list.Add(new GenericParameterHelper(1));
+        using (RarelyRemoveItemSet<GenericParameterHelper>.Enumerator enumerator = this.list.EnumerateAndClear().GetEnumerator())
+        {
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(1, enumerator.Current.Data);
+            Assert.False(enumerator.MoveNext());
+            enumerator.Reset();
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(1, enumerator.Current.Data);
+            Assert.False(enumerator.MoveNext());
+        }
+    }
+
+    [Fact]
+    public void EnumerationOfTwo()
+    {
+        this.list.Add(new GenericParameterHelper(1));
+        this.list.Add(new GenericParameterHelper(2));
+        using (RarelyRemoveItemSet<GenericParameterHelper>.Enumerator enumerator = this.list.EnumerateAndClear().GetEnumerator())
+        {
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(1, enumerator.Current.Data);
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(2, enumerator.Current.Data);
+            Assert.False(enumerator.MoveNext());
+            enumerator.Reset();
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(1, enumerator.Current.Data);
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal<int>(2, enumerator.Current.Data);
+            Assert.False(enumerator.MoveNext());
+        }
+    }
+
+    [Fact]
+    public void RemoveFromEmpty()
+    {
+        this.list.Remove(null!);
+        Assert.Empty(this.list.ToArray());
+        this.list.Remove(new GenericParameterHelper(5));
+        Assert.Empty(this.list.ToArray());
+    }
+
+    [Fact]
+    public void RemoveFromOne()
+    {
+        var value = new GenericParameterHelper(1);
+        this.list.Add(value);
+
+        this.list.Remove(null!);
+        Assert.Single(this.list.ToArray());
+        this.list.Remove(new GenericParameterHelper(5));
+        Assert.Single(this.list.ToArray());
+        this.list.Remove(value);
+        Assert.Empty(this.list.ToArray());
+    }
+
+    [Fact]
+    public void RemoveFromTwoLIFO()
+    {
+        var value1 = new GenericParameterHelper(1);
+        var value2 = new GenericParameterHelper(2);
+        this.list.Add(value1);
+        this.list.Add(value2);
+
+        this.list.Remove(null!);
+        Assert.Equal(2, this.list.ToArray().Length);
+        this.list.Remove(new GenericParameterHelper(5));
+        Assert.Equal(2, this.list.ToArray().Length);
+        this.list.Remove(value2);
+        Assert.Single(this.list.ToArray());
+        Assert.Equal(1, this.list.ToArray()[0].Data);
+        this.list.Remove(value1);
+        Assert.Empty(this.list.ToArray());
+    }
+
+    [Fact]
+    public void RemoveFromTwoFIFO()
```

Comment:
Given your technique of *not* shifting all items to the left upon removal, I feel you should have at least one test with 4 items and you remove the 2nd item so that we can assert that 1st, 3rd and 4th position items remain accessible to enumeration.